File: 프로그래머스/2/176962. 과제 진행하기/과제 진행하기.py
def solution(plans):
    # 시작 시간을 분으로 변경하고 오름차순으로 정렬
    for i in range(len(plans)):
        hour, minute = plans[i][1].split(":")
        plans[i][1] = int(hour) * 60 + int(minute)
        plans[i][2] = int(plans[i][2])
    plans.sort(key=lambda x: x[1])

    stack = []
    result = []
    for i in range(len(plans)-1):
        cur, next = plans[i], plans[i+1]
        remaining_time = next[1] - cur[1]
        if remaining_time < cur[2]:
            cur[2] -= remaining_time
            stack.append(cur)
        else:
            result.append(cur[0])
            remaining_time -= cur[2]

            while stack:
                if stack[-1][2] > remaining_time:
                    stack[-1][2] -= remaining_time
                    break
                else:
                    remaining_time -= stack[-1][2]
                    v = stack.pop()
                    result.append(v[0])
    result.append(plans[-1][0])

    if stack:
        stack = [v[0] for v in stack]
        stack = stack[::-1]
        result += stack

    return result

File: 프로그래머스/2/176962. 과제 진행하기/과제 진행하기.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Stack;

class Solution {
    public String[] solution(String[][] plans) {
        Arrays.sort(plans, (o1, o2) -> {
            return getTime(o1[1]) - getTime(o2[1]);
        });

        Stack<Tuple> st = new Stack<>();
        List<String> answer = new ArrayList<>();

        for (int i = 0; i < plans.length - 1; i++) {
            String[] plan = plans[i];
            String name = plan[0];
            System.out.println("name: " + name);
            int time = getTime(plan[1]);
            int remain = Integer.parseInt(plan[2]);

            String[] nextPlan = plans[i + 1];
            int nextTime = getTime(nextPlan[1]);

            if (time + remain > nextTime) {
                st.add(new Tuple(time + remain - nextTime, name));
                continue;
            }

            int v = nextTime - time - remain;
            answer.add(name);

            while (v > 0 && !st.isEmpty()) {
                Tuple tuple = st.pop();
                if (tuple.remain > v) {
                    st.push(new Tuple(tuple.remain - v, tuple.name));
                    break;
                } else {
                    v -= tuple.remain;
                    answer.add(tuple.name);
                }
            }
        }

        answer.add(plans[plans.length-1][0]);

        while (!st.isEmpty()) {
            Tuple tuple = st.pop();
            answer.add(tuple.name);
        }


        return answer.toArray(String[]::new);
    }

    public int getTime(String time) {
        String[] split = time.split(":");
        int h = Integer.parseInt(split[0]);
        int m = Integer.parseInt(split[1]);

        return h * 60 + m;
    }
}

class Tuple {

    public int remain;
    public String name;

    public Tuple(int remain, String name) {
        this.remain = remain;
        this.name = name;
    }
}